src/persist/src/operators/stream.rs

```diff
@@ -41,46 +51,252 @@ where
+    /// this operator is holding on to capabilities as long as seals corresponding to their
+    /// timestamp are not done.
+    fn seal(
+        &self,
+        name: &str,
+        write: StreamWriteHandle<K, V>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    );
+}
+
+impl<G, K, V> Seal<G, K, V> for Stream<G, ((K, V), u64, isize)>
+where
+    G: Scope<Timestamp = u64>,
+    K: TimelyData + Codec,
+    V: TimelyData + Codec,
+{
+    fn seal(
+        &self,
+        name: &str,
+        write: StreamWriteHandle<K, V>,
+    ) -> (
+        Stream<G, ((K, V), u64, isize)>,
+        Stream<G, (String, u64, isize)>,
+    ) {
+        let operator_name = format!("seal({})", name);
+        let mut seal_op = OperatorBuilder::new(operator_name.clone(), self.scope());
+
+        let mut data_input = seal_op.new_input(&self, Pipeline);
+
+        let (mut data_output, data_output_stream) = seal_op.new_output();
+        let (mut error_output, error_output_stream) = seal_op.new_output();
+
+        let mut data_buffer = Vec::new();
+        let mut input_frontier =
+            Antichain::from_elem(<G::Timestamp as timely::progress::Timestamp>::minimum());
+        let mut capabilities = Antichain::<Capability<u64>>::new();
+
+        // We only seal from one worker because sealing from multiple workers could lead to a race
+        // conditions where one worker seals up to time `t` while another worker is still trying to
+        // write data with timestamps that are not beyond `t`.
+        //
+        // Upstream persist() operators will only advance their frontier when writes are succesful.
+        // With timely progress tracking we are therefore sure that when the frontier advances for
+        // worker 0, it has advanced to at least that point for all upstream operators.
+        //
+        // Alternative solutions would be to "teach" persistence to work with seals from multiple
+        // workers, or to use a non-timely solution for keeping track of outstanding write
+        // capabilities.
+        let active_seal_operator = self.scope().index() == 0;
+
+        seal_op.build(move |_capabilities| {
+            move |frontiers| {
+                let mut data_output = data_output.activate();
+                let mut error_output = error_output.activate();
+
+                // Pass through all data.
+                data_input.for_each(|cap, data| {
+                    data.swap(&mut data_buffer);
+
+                    let mut session = data_output.session(&cap);
+                    session.give_vec(&mut data_buffer);
+
+                    // We only need capabilities for reporting errors, which we only need to do
+                    // when we're the active operator.
+                    if active_seal_operator {
+                        capabilities.insert(cap.retain_for_output(1));
+                    }
+                });
+
+                if active_seal_operator {
```

Comment:
same nit: can we invert the control flow here to be `if !active_seal_operator { return }`?